### ISSUE TITLE
[ScatterElements] Clarify the requirement on indices uniqueness

### ISSUE THREAD

**Issue by masahi** [onnx/onnx#3234]
[The spec](https://github.com/onnx/onnx/blob/master/docs/Operators.md#ScatterElements) of `ScatterElements` op doesn't tell if elements of `indices` tensor need to be unique or not along the scattered axis. I think there should be an explicit clarification on this, because it would determine if non-deterministic outputs from "embarassingly parallel" scatter implementation (e.g. on GPU) are allowed or not.

Here some data points for reference:
* [`ScatterND` op spec](https://github.com/onnx/onnx/blob/master/docs/Operators.md#ScatterElements) does say indices must be unique. 
* [The PyTorch doc](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.scatter_) says indices must be unique. 
* [ONNXRuntime implementation of scatter op ](https://github.com/microsoft/onnxruntime/blob/bba185a582655b95fdd9913ac45f3f5c8ffd10ba/onnxruntime/core/providers/cuda/tensor/scatter_elements_impl.cu#L174-L217) clearly does scattering in parallel, so it is making an implicit assumption that indices are unique.

**Comment by masahi** [comment]
cc @askhade @jcwchen What do you think?

**Comment by jcwchen** [comment]
Hi @masahi,
Thank you providing the details. I think it makes sense to me. Could someone in @onnx/sig-operators verify this? Thanks.

**Comment by gramalingam** [comment]
I agree (personally speaking). I think an equivalent way of saying it is that the spec does not guarantee a specific execution ordering for the updates (and so a unique output is guaranteed only if there indices are unique). The question is: is there any user (model) that is assuming otherwise and has a dependence on it? It seems unlikely, but would be good to verify. Does TF have the same assumption?

**Comment by masahi** [comment]
TF also says the output is non deterministic if there duplicated indices https://www.tensorflow.org/api_docs/python/tf/compat/v1/scatter_update

JAX too, https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scatter.html

So the industry clearly favors more performance than guaranteeing determinism. To support modes, we can add a new attribute to specify if indices are supposed to be unique or not. That's what I'm thinking I'll do for [TVM](https://github.com/apache/tvm), which currently guarantees determinism and hence slow.

**Comment by postrational** [comment]
I think we should avoid creating a new version of the operator, which acts a a superset of both behaviours. This would place the burden on implementors, to support both modes.

If the industry favours the parallel, non-deterministic approach, than we should document that this is what ONNX expects as well. If we do run into models, where this causes problems, we can come back to this issue and reconsider.

We should update our docs to be more explicit about the behaviour, but without introducing a new operator version.

**Comment by gramalingam** [comment]
I also agree. Let us clarify the specification, not add an attribute to support both mode now. We can add it if and when there is a need for the second mode.

**Comment by jcwchen** [comment]
Reopen it since this issue hasn't been resolved yet.
